overwallclock fix

dbeltrankyl opened the pull request:
Fixes:

* This aims to fix the issue with the 4.1.12-dev not being able to cancel jobs that end by wallclock. 

Issues detected:

* (is_overwallclock) Depending on what was set in the wallclock ( 02:00 (hours) or 00:20 (minutes) it was wrongly multiplying the value * 60 
* timeout was always being multiplying by 60

Work done:

* Added a few TODO to be done in another branch so this is not too big and is not related to the main issue
* Removed some duplicated functions and merged them to the job class.
* Changed the self.is_overwallclock call 
* Changed the parse_time call to not give the format anymore as it is not needed and leads to bugs
* Added self.wallclock_in_seconds property 
* Added few warnings
* Added that in case that the wallclock *1.3 > max_wallclock, set the wallclock to the max_wallclock

Todo:

* [  ] Pytests

kinow commented on autosubmit/platforms/psplatform.py:
```diff
@@ -96,13 +96,11 @@ def jobs_in_queue(self):
         else:
             export += " ; "
-        return self.get_call(job_script, job, export = export, timeout=seconds)
+        return self.get_call(job_script, job, export=export, timeout=job.wallclock_in_seconds)
```
Good idea to move it to the Job :+1:

kinow commented on autosubmit/job/job.py:
```diff
@@ -1264,49 +1283,61 @@ def retrieve_logfiles(self, platform: Any, raise_error: bool = False) -> Dict[st
-            wallclock_delta = datetime.timedelta(minutes=minute, seconds=second)
-        if elapsed > wallclock_delta:
+        elapsed = datetime.datetime.now() - self.start_time
+        if int(elapsed.total_seconds()) > self.wallclock_in_seconds:
```
Not sure if this works as expected?

```py
>>> import datetime
>>> 
>>> start_time = datetime.datetime.now()
>>> 
>>> elapsed = datetime.datetime.now() - start_time
>>> 
>>> print(elapsed)
0:00:11.001284
>>> type(elapsed)
<class 'datetime.timedelta'>
>>> int(elapsed)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: int() argument must be a string, a bytes-like object or a real number, not 'datetime.timedelta'
```

I think it would result in a `TypeError` here?

dbeltrankyl replied:
erm, not sure why it is not working in the console

The start_time value didn't change, before it was sending the job.start_time and now it directly calls to self.start_time but the value is the same as before

start_time == self.start_time == time.time()

dbeltrankyl replied:
nvm, 

the start_time that is being recieved is 

# paramiko_platform.py line 647-651

```python
            elif job_status in self.job_status['RUNNING']:
                job_status = Status.RUNNING
                if not is_wrapper:
                    if job.status != Status.RUNNING:
                        job.start_time = datetime.datetime.now() # URi: start time
```

dbeltrankyl replied:
I believe that

job.py line 2083 - 2084
```python
        if hasattr(self, "start_time"):
            self.start_time = time.time()
```
should be

```python
        if not hasattr(self, "start_time"):
            self.start_time = datetime.datetime.now()
```

dbeltrankyl replied:
changed

kinow reviewed (commented):
Without tests I think it'd need some more thorough review. I started reading parts of the change (really liked moving the logic of dates for jobs to the `Job` class, seems to clean the code a bit :1st_place_medal:), but reading it quickly I found at least one place that looks like it could raise a `TypeError` (see comment).

dbeltrankyl commented:
Hello @kinow 

I did a manual test ( was preparing a yaml file to pass to you) with wrappers and it fails because the wallclock is not stored in the wrappers db.. preparing a fix 

Then I'll revise the feedback you mentioned

dbeltrankyl commented:
Also, I think it doesn't only affect this, but also the is_wrapper value. 

I believe that experiments with wrappers wouldn't be able to do: 

module load autosubmit/4.1.11
run $expid
module load autosubmit/4.1.12
run $expid -v

dbeltrankyl commented:
I don't have time to write the pytests ( in fact, I'm already in over hours )

Neither the changelog ( apologies for that, this issue was unexpected)

@kinow , I propose two options: 

* Deploy the beta version with this without pytests ( after a manual check )

To test it 

* minimal.yml with project_type to none

* test.yml
```yaml
CONFIG:
  # Current version of autosubmit.
  AUTOSUBMIT_VERSION: "4.1.12"
  # Maximum number of jobs permitted in the waiting status.
  MAXWAITINGJOBS: 20
  # Total number of jobs in the workflow.
  TOTALJOBS: 20
  SAFETYSLEEPTIME: 0
MAIL:
  NOTIFICATIONS: False
  TO:
STORAGE:
  TYPE: pkl
  COPY_REMOTE_LOGS: true

EXPERIMENT:
  DATELIST: 20221101
  MEMBERS: fc0
  CHUNKSIZEUNIT: month
  CHUNKSIZE: 2
  NUMCHUNKS: 3
  CHUNKINI: ''
  CALENDAR: standard
PROJECT:
  # Type of the project.
  PROJECT_TYPE: none
  # Folder to hold the project sources.
  PROJECT_DESTINATION: git_project
GIT:
  PROJECT_ORIGIN: "blabla"
  PROJECT_BRANCH: "blabla"
  PROJECT_COMMIT: ''
  PROJECT_SUBMODULES: ''
  FETCH_SINGLE_BRANCH: true
PLATFORMS:
  MARENOSTRUM5:
    TYPE: slurm
    HOST: glogin1.bsc.es
    PROJECT: bsc32
    USER: "FILL"
    QUEUE: gp_debug
    SCRATCH_DIR: /gpfs/scratch
    ADD_PROJECT_TO_HOST: false
    MAX_WALLCLOCK: 02:00
    TEMP_DIR: ''
    MAX_PROCESSORS: 99999
    PROCESSORS_PER_NODE: 112
  MARENOSTRUM5-login:
    TYPE: ps
    HOST: glogin2.bsc.es
    PROJECT: bsc32
    USER: "FILL"
    SCRATCH_DIR: /gpfs/scratch
    ADD_PROJECT_TO_HOST: false
    MAX_WALLCLOCK: 02:00
    TEMP_DIR: ''
    MAX_PROCESSORS: 99999
  
JOBS:
  mn5-ps:
    SCRIPT: |
      sleep 180
      echo "Hello World with id=Success" # this is not in the log. # Complete file doesn't exists.
    RUNNING: once
    wallclock: 00:01 # should be 78 seconds
    PLATFORM: MARENOSTRUM5-login

  mn5-slurm:
    SCRIPT: |
      sleep 180
      echo "Hello World with id=Success" # this is not in the log.
    RUNNING: once
    wallclock: 00:01 # should be 78 seconds  == self._wallclock_in_seconds == 78s
    PLATFORM: MARENOSTRUM5

  local:
    SCRIPT: |
      sleep 180
      echo "Hello World with id=Success" # this is not in the log.
    RUNNING: once
    wallclock: 00:01 # should be 78 seconds  == self._wallclock_in_seconds == 78s
    PLATFORM: LOCAL

  mn5wrapper:
    SCRIPT: |
      sleep 180 
      echo "Hello World with id=Success" # this is not in the log.
    DEPENDENCIES:
      mn5wrapper-1:
    RUNNING: chunk
    wallclock: 00:01 # should be 78(seconds)*·3(chunks) == self._wallclock_in_seconds == 234s
    PLATFORM: MARENOSTRUM5

wrappers:
  mn5wrapper:
    JOBS_IN_WRAPPER: mn5wrapper
    TYPE: vertical
```


* Another person writes the pytest for this and then merges it

ABOUT THE WRAPPERS: 

There was an additional issue: whenever I stopped and rerun with an auto-submit run -> auto-submit run. The wallclock value was not stored, so the self._wallclock_in_seconds was reset to 0.

So, to test that part, you need to stop the experiment once the wrapper job is running and then run again.

There was another issue related to the packed value not being reset on a reload, so the preview (-cw) was not working well after a wrapper was run once.

dbeltrankyl commented:
Btw I think that the wrapper issue could be related to https://github.com/BSC-ES/autosubmit/issues/2027

dbeltrankyl commented:
> Also, I think it doesn't only affect this, but also the is_wrapper value.
> 
> I believe that experiments with wrappers wouldn't be able to do:
> 
> module load autosubmit/4.1.11 run $expid module load autosubmit/4.1.12 run $expid -v

About this, I had to add a new field to the wrapper database.

This database is re-created on each run, so the only thing not compatible is the run -> run -v.

I've also left comments in the last commits

dbeltrankyl commented:
Hello @kinow 

Did you review this one? https://github.com/BSC-ES/autosubmit/pull/2041#issuecomment-2551779623 . I did a question there

kinow commented:
Not yet, @dbeltrankyl . With this one the failure happened in the testing suite only because of slowness in MN5's GPFS, and the experiment stopped. The main issue was the remote task continued running (sometimes to completion depending on the remote wallclock, I think?). So this issue seemed to have lower priority than the one in autosubmit-config-parser for @franra9 's apps workflows.

kinow commented on autosubmit/job/job.py:
```diff
@@ -273,6 +274,9 @@ def _adjust_new_parameters(self) -> None:
                 self.wrapper_name = "wrapped"
+        if not hasattr(self, '_wallclock_in_seconds'):  # Added in 4.1.12
+            self._wallclock_in_seconds = None
+            self.wallclock_in_seconds = self.wallclock
```
If I create a `Job` object, the `wallclock_in_seconds` will be `None`:

```py
(autosubmit4) kinow@ranma:~/Development/python/workspace/autosubmit$ python3
Python 3.9.0 | packaged by conda-forge | (default, Nov 26 2020, 07:57:39) 
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from autosubmit.job.job import Job
>>> j = Job('a', 'a', 'a', 1)
>>> j.wallclock_in_seconds
```

Then, if I add a wallclock, then the `job` object is still in some weird state, where it's a Job with a wallclock, but without a wallclock in seconds.

```py
>>> j.wallclock = '01:00'
>>> j.wallclock_in_seconds
```

Then, if I call `adjust_parameters()`, `wallclock_in_seconds` is still `None`, because `hasattr(self, '_wallclock_in_seconds')` returns `True`.

```py
>>> j._adjust_new_parameters()
>>> j.wallclock_in_seconds
```

The only way to have that property is calling is_overwallclock first, I think. And then you get the time in seconds `* 1.30`.

```py
>>> import datetime
>>> j.start_time = datetime.datetime.now()
>>> j.is_over_wallclock()
False
>>> j.wallclock_in_seconds
4680
```

I thought about suggesting to have just a getter `@property`, but I think the reason why you used a variable to persist that state was to avoid recalculating it for each job... It's just that it looks a bit odd & risky the possible states in which a  `Job` will be, and having to call `is_over_wallclock` before being able to use `wallclock_in_seconds`.

dbeltrankyl replied:
So, what if whenever the wallclock is set , it also sets the wallclock_in_seconds? 

And change

```
        if not hasattr(self, '_wallclock_in_seconds'):  # Added in 4.1.12
for
        if not hasattr(self, '_wallclock_in_seconds') or not self.wallclock_in_seconds:  # Added in 4.1.12
```

kinow replied:
I think this has been fixed in the latest changes. Now both `_wallclock` and `_wallclock_in_seconds` (and their properties) should all be in sync :clap: marking as resolved. Thanks!

kinow reviewed (commented):
I had left a few files pending review the last time, and stopped after sending the last feedback.

This time I finished reading the whole code. Left some comments/questions, but the branch also seems to have a Git conflict and the coverage report (last comment of the bot, at least) is pointing to some parts of the code not tested, @dbeltrankyl .

Normally that's OK, but if it's not too hard to test those lines too, then it'd reduce the risk for merging it prior to the d-suite :muscle: 

Cheers

dbeltrankyl commented on autosubmit/autosubmit.py:
```diff
@@ -2059,12 +2061,17 @@ def prepare_run(expid, notransitive=False, start_time=None, start_after=None,
             Log.debug("Processing job packages")
             try:
-                for (exp_id, package_name, job_name) in packages:
+                wrapper_wallclock = "48:00" # fallback value, only affects to is_overclock
```
done, added doc

dbeltrankyl commented on autosubmit/autosubmit.py:
```diff
@@ -2059,12 +2061,17 @@ def prepare_run(expid, notransitive=False, start_time=None, start_after=None,
+            except ValueError as e:  # New field in the db, so not compatible if the wrapper package is not reset ( done in the create function )
+                if "not enough values to unpack" in str(e).lower():
+                    raise AutosubmitCritical("Error while loading the wrappers, this due a AS version change in the middle of the run. Please, run 'autosubmit create -f' again.", 7014, str(e))
+                raise
```
The outcome would be similar. Both stuff are gathered by bin/autosubmit and show the traceback with a slightly different message

But your solution is cleaner, so I've added it. Thanks!

dbeltrankyl commented on autosubmit/autosubmit.py:
```diff
@@ -5754,6 +5761,9 @@ def load_job_list(expid, as_conf, notransitive=False, monitor=False, new = True)
         else:
             job_list.remove_rerun_only_jobs(notransitive)
 
+        for job in job_list.get_waiting() + job_list.get_ready(): # Reset packed value on load so the jobs can be wrapped again. Inspect -cw and Create -cw had issues.
```
I think the warning is due to a missing space, but I moved it

dbeltrankyl commented on autosubmit/job/job.py:
```diff
@@ -292,11 +296,26 @@ def _init_runtime_parameters(self):
+        if not self._wallclock_in_seconds or self.status not in [Status.RUNNING, Status.QUEUING, Status.SUBMITTED]:
+            # Should always take the max_wallclock set in the platform, this is set as fallback ( and local platform doesn't have a max_wallclock defined )
+            if not wallclock or wallclock == "00:00":
+                wallclock = "24:00"
```
For the sysadmin it doesn't make sense because it depends on the platform, so we can't add it to the autosubmitrc file. 

Added it in the CONFIG section, added doc

dbeltrankyl commented on autosubmit/job/job.py:
```diff
@@ -1264,49 +1283,63 @@ def retrieve_logfiles(self, platform: Any, raise_error: bool = False) -> Dict[st
+                return None
+            return int(wallclock.total_seconds())
+        else:
+            return None
```
applied, thanks!

dbeltrankyl commented on autosubmit/job/job.py:
```diff
@@ -292,11 +296,26 @@ def _init_runtime_parameters(self):
+                wallclock = "24:00"
+                Log.warning(f"No wallclock is set for this job. Default to {wallclock}")
+            wallclock = self.parse_time(wallclock)
+            self._wallclock_in_seconds = self._time_in_seconds_and_margin(wallclock)
```
I think this should be implemented in the config parser in a verification function ( only if all files are found and loaded, or the proj/ folder exists in advanced configurations that that is the tricky part) )

kinow replied:
Issue: https://github.com/BSC-ES/autosubmit-config-parser/issues/65

Resolving.

dbeltrankyl commented on autosubmit/job/job.py:
```diff
@@ -1264,49 +1283,63 @@ def retrieve_logfiles(self, platform: Any, raise_error: bool = False) -> Dict[st
-    def parse_time(self,wallclock):
+    def _max_possible_wallclock(self):
+        if self.platform and self.platform.max_wallclock:
+            wallclock = self.parse_time(self.platform.max_wallclock)
```
Same as previous comment, Shouldn't this be handled in the Autosubmit config parser? in some verification step. The value here should be expected to be correct, no?

kinow replied:
I would expect so, to be validated in one of the two places. I will test again with the latest changes, and if it's still happening then I'll create an issue there in the config-parser (we can move it to another repo like this one later if needed). Thanks!

kinow replied:
Done: https://github.com/BSC-ES/autosubmit-config-parser/issues/65

Will mark both discussions as resolved. Thanks Dani!

dbeltrankyl commented on autosubmit/job/job.py:
```diff
@@ -1264,49 +1283,63 @@ def retrieve_logfiles(self, platform: Any, raise_error: bool = False) -> Dict[st
+
+        This method increases the given wallclock time by 30%.
+        It then converts the total wallclock time to seconds and returns both the total
+        wallclock time in seconds and the wallclock time with the margin as a timedelta.
```
> This is only for ps platforms, right? Slurm platform will handle the wallclock in another way.

The buffer is for all platforms, but SLURM/PJM etc, will cap at the max_wallclock possible.

> Does it make sense to use 30% for every platform (Fugaku, KIT, EuroHPC, etc.)?

I believe yes, this is just if the template took longer than the user expected with the wallclock so there is not difference with the platform used. 

 > Why do we use a buffer of 30% here? Or should it be configurable?

I tried to find the issue but I can't find it. If I recall correctly, this was recommended by Miguel some years ago. Not sure if some calculation was involved or was just a "lets put this number" case.  

 In line with the previous quote, some sudden issues in the platforms (some core performing slower, filesystem issues...) can cause the job to run slower and thus hit the wallclock.

This was set in a time when the mn4 was performing badly. 

I think we can it keep as 30% and move this discussion to a new issue

kinow replied:
Thanks for the explanation, @dbeltrankyl ! Moved discussion to https://github.com/BSC-ES/autosubmit/issues/2075, marking as resolved.

dbeltrankyl commented:
Rebased, feedback commented. Waiting for coverage and write test

dbeltrankyl commented:
Hello @kinow 

I've added some tests.

I had to rework the function that checks the status of the wrapper after a load because it was not right ( discovered after doing the test to test the wallclock part)

I also added a function to properly test the overclock function. There is uncovered stuff related to the calls to that function because it is in a spot hard to easily test

kinow reviewed (commented):
@dbeltrankyl I think there are some conflicts in Git here.. can you have a look whenever you have time, please?

dbeltrankyl commented:
Done

kinow reviewed (commented):
14 out of 20 files reviewed. Will try to finish it tomorrow. Submitting the review of what I have now. Thanks @dbeltrankyl for adding new configs, adjusting the code, adding docs, and for the many new tests you added over past days!

kinow commented on autosubmit/job/job_package_persistence.py:
```diff
@@ -35,51 +37,57 @@ class JobPackagePersistence(object):
+                if len(wrapper) != 4:
+                    # New field in the db, so not compatible if the wrapper package is not reset (done in the create function)
+                    raise AutosubmitCritical("Error while loading the wrappers, the current wrappers has a different amount of fields than the expected."
+                                             "This due a AS version change without doing autosubmit create -f. "
```
Please, run" needs a space, Please, run "

dbeltrankyl replied:
Done thanks!

dbeltrankyl commented on test/unit/test_paramiko_platform.py:
```diff
@@ -145,3 +145,21 @@ def test_send_file(mocker, ps_platform, filename, check):
+    job.section = 'dummysection'
+    job.platform_name = 'pytest-ps'
+    job.platform = platform
+    job.wallclock_in_seconds = 60
```
Hello @kinow 

Thanks for the feedback, and the patches 

I don't get this point: 

+    job.wallclock_in_seconds = '00:00:60'

We don't want the wallclock_in_seconds being in that format. It has to be a int so we can add it to the 

timeout <int> <submit.cmd>

dbeltrankyl replied:
about the rest I agree is what I ask there  https://github.com/BSC-ES/autosubmit/pull/2041#discussion_r1916330515. thanks!

dbeltrankyl commented on autosubmit/platforms/paramiko_platform.py:
```diff
@@ -586,6 +586,21 @@ def parse_estimated_time(self, output):
+                job_status = Status.FAILED
+
+            if cancel and job_status is Status.FAILED:
+                with suppress(Exception):
```
Added

dbeltrankyl commented:
Thanks Bruno! 

Feedback addressed I believe

kinow commented:
Ok, resolved all discussions here, synced the code, and finished replying all my pending notifications (slack/email/github/mattermost). Now will eat something, and then focus on the pending code to review here, before a final test (with a simple experiment + ps platform and create+inspect to confirm before I got wallclock*60, with the branch I get the wallclock seconds as expected).

dbeltrankyl commented on test/unit/test_overwallclock.py:
```diff
@@ -0,0 +1,166 @@
+    platform_instance.cancel_cmd = None
+    platform_instance.send_command = mocker.MagicMock()
+    job_status = platform_instance.job_is_over_wallclock(job, Status.RUNNING, True)
+    platform_instance.send_command.assert_not_called()
```
It was intentional, but I've removed it. I only wanted to check if the command was called in that part.

dbeltrankyl commented:
Thanks @kinow 

Feedback resolved

kinow commented on autosubmit/job/job.py:
```diff
@@ -37,7 +37,7 @@
 from threading import Thread
 from time import sleep
-from typing import List, Union, Dict, Any
+from typing import List, Union, Dict, Any, Tuple
```
I think this one is still not used :point_up: , @dbeltrankyl

kinow commented:
Left one question about an unused import, but didn't find anything else to comment in the code :tada: going to run some autosubmit create + inspect now, to verify the issue is indeed fixed, and check coverage. Thanks Dani!

kinow commented:
Using master first.

Created a dummy experiment, changed to project type, set some dummy scripts, a ps platform pointing to localhost, and then used wallclock of 02:00 in LOCAL_SETUP.

Created and inspected, all good, then run shows the timeout command in aslogs run log.

```bash
2025-01-23 09:16:38,619 Submitting job with the command: timeout 432000 nohup bash /tmp/bdepaula/a00e/LOG_a00e/a00e_LOCAL_SETUP.cmd > /tmp/bdepaula/a00e/LOG_a00e/a00e_LOCAL_SETUP.cmd.out.0 2> /tmp/bdepaula/a00e/LOG_a00e/a00e_LOCAL_SETUP.cmd.err.0 & echo $!
2025-01-23 09:16:38,621 Job: a00e_LOCAL_SETUP submitted with job_id: 12637
```

So two hours became 432000 seconds, instead of 7200.

Using this branch.

I tried `create` & `inspect` without the `-v`, and got

```bash
[WARNING] Structures folder not found /home/bdepaula/autosubmit/metadata/structures
Looking for edgeless jobs...
Starting to generate cmd scripts
[WARNING] There is a .lock file and not -f, generating only all unsubmitted cmd scripts
[WARNING] Generating the auxiliary job_list used for the -CW flag.
[WARNING] Aux Job_list was generated successfully
Calculating possible ready jobs for FAKE-LOCAL
 [CRITICAL] Unexpected error: 'Job' object has no attribute '_wallclock_in_seconds'.
 Please report it to Autosubmit Developers through Git
More info at https://autosubmit.readthedocs.io/en/master/troubleshooting/error-codes.html
```

I think users would have to `autosubmit create -f`. I think my feedback from yesterday didn't quite work -- your code had a protection against this in ... adjust parameters I think? I think we will have to have at least a `self._wallclock_in_seconds = None` there...

Did an `autosubmit create -f`, `inspect`, and then `run`.

This time I got quite close to what I expected, but exactly what's expected:

```bash
2025-01-23 09:29:08,526 Submitting job with the command: timeout 9360 nohup bash /tmp/bdepaula/a00e/LOG_a00e/a00e_LOCAL_SETUP.cmd > /tmp/bdepaula/a00e/LOG_a00e/a00e_LOCAL_SETUP.cmd.out.0 2> /tmp/bdepaula/a00e/LOG_a00e/a00e_LOCAL_SETUP.cmd.err.0 & echo $!
2025-01-23 09:29:08,528 Job: a00e_LOCAL_SETUP submitted with job_id: 13817
```

I expected 7200, but in reality we add 30% more in Autosubmit. So 7200 * 1.3, 9360. :ok_hand:

dbeltrankyl commented:
> I think users would have to autosubmit create -f. I think my feedback from yesterday didn't quite work -- your code had a protection against this in ... adjust parameters I think? I think we will have to have at least a self._wallclock_in_seconds = None there...

```yaml
 
        if not hasattr(self, "_wallclock_in_seconds"):  # Added in 4.1.12
            self._wallclock_in_seconds = None
            self.wallclock = self.wallclock # also sets the wallclock in seconds

```

Added this, the self.wallclock = self.wallclock is to set the wallclock_in_seconds to the expected value and the first none is because if not the assign complains that the wallclock_in_seconds doesn't exists.

kinow reviewed (approved):
Lines not covered do not appear to be critical. Code looks good, tested and worked well. New code looks good and appears to fix the issue when the variable does not exist when loaded from job_list/disk. Great job, Dani!
